### PR TITLE
Qbft RoundChangeMessage validator creation

### PR DESCRIPTION
**Pull request by rain-on**
## PR description
This supplies the necessary validation logic for parsing received RoundChange messages.
This logic is wired into the QBFT business logic.
## Fixed Issue(s)
N/A

## Changelog
N/A
- [X] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangeMessageValidator.java [hunk 1]
```diff
@@ -14,11 +14,135 @@
...
+
+    if (msg.getPreparedRoundMetadata().isEmpty()) {
+      LOG.info("{}: Prepared block specified, but prepared metadata absent", ERROR_PREFIX);
+      return false;
+    }
+
+    final PreparedRoundMetadata metadata = msg.getPreparedRoundMetadata().get();
+
+    if (!metadata.getPreparedBlockHash().equals(block.getHash())) {
+      LOG.info("Invalid RoundChange message: Prepared metadata hash does not match supplied block");
```
Comment (usmansaleem) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangeMessageValidator.java:110]: nit: Use ERROR_PREFIX as utilized in other logging messages.

**Review by usmansaleem — APPROVED** [review]
LGTM - one very minor/nit-pick enhancement suggested.

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangeMessageValidator.java (hunk 1 above)
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangeMessageValidator.java:110]: done

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangeMessageValidator.java [hunk 2]
```diff
@@ -14,11 +14,133 @@
...
+
+    if (msg.getPreparedRoundMetadata().isEmpty()) {
+      LOG.info("{}: Prepared block specified, but prepared metadata absent", ERROR_PREFIX);
+      return false;
+    }
+
+    final PreparedRoundMetadata metadata = msg.getPreparedRoundMetadata().get();
+
+    if (!metadata.getPreparedBlockHash().equals(block.getHash())) {
+      LOG.info("{}: Prepared metadata hash does not match supplied block");
```
Comment (jframe) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangeMessageValidator.java:110]: Log statement is missing an ERROR_PREFIX arg

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangeMessageValidator.java [hunk 3]
```diff
@@ -14,11 +14,133 @@
...
+  private boolean validatePrepares(
+      final PreparedRoundMetadata metaData, final List<SignedData<PreparePayload>> prepares) {
+
+    final ConsensusRoundIdentifier preparedRoundIdentifier =
+        new ConsensusRoundIdentifier(chainHeight, metaData.getPreparedRound());
+
+    final PrepareValidator validator =
+        new PrepareValidator(validators, preparedRoundIdentifier, metaData.getPreparedBlockHash());
+
+    if (hasDuplicateAuthors(prepares)) {
```
Comment (jframe) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangeMessageValidator.java:134]: I think duplicate messages are allowed, they just need to be counted as one when determining quorum.

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangeMessageValidator.java [hunk 4]
```diff
@@ -14,11 +14,133 @@
...
+    this.quorumMessageCount = quorumMessageCount;
+    this.chainHeight = chainHeight;
+    this.validators = validators;
+    this.blockValidator = blockValidator;
+    this.protocolContext = protocolContext;
+  }
+
+  public boolean validate(final RoundChange msg) {
+
+    if (!roundChangePayloadValidator.validate(msg.getSignedPayload())) {
```
Comment (jframe) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangeMessageValidator.java:70]: height should be checked that it matches as well

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/ValidationHelpers.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+ */
+package org.hyperledger.besu.consensus.qbft.validation;
+
+import org.hyperledger.besu.consensus.common.bft.payload.Authored;
+import org.hyperledger.besu.consensus.common.bft.payload.Payload;
+import org.hyperledger.besu.consensus.common.bft.payload.SignedData;
+
+import java.util.Collection;
+
+public class ValidationHelpers {
```
Comment (jframe) [inline consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/ValidationHelpers.java:23]: This isn't needed anymore now that there is a common.bft ValidationHelpers

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangeMessageValidator.java (hunk 2 above)
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangeMessageValidator.java:110]: done

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangeMessageValidator.java (hunk 4 above)
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/RoundChangeMessageValidator.java:70]: This check exists in the RoundChangePayloadValidator.

File: consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/ValidationHelpers.java (hunk 5 above)
Comment (rain-on) [reply consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/validation/ValidationHelpers.java:23]: done

**Pull request merged**